spec/views/components/autocomplete_spec.rb

```diff
@@ -0,0 +1,90 @@
+require 'rails_helper'
+
+RSpec.describe "Autocomplete component" do
+  it "should not render when no data is given" do
+    assert_empty render_component({})
+  end
+
+  it "should not render when only a name is provided" do
+    assert_empty render_component(name: "foo")
+  end
+
+  it "should not render when only a label is provided" do
+    assert_empty render_component(label: { text: "foo" })
+  end
+
+  it "should render text input with randomised id when name & label provided" do
+    render_component(name: "foo", label: { text: "bar" })
+
+    assert_select ".app-c-autocomplete", count: 1
+    randomised_for = nil
+    assert_select ".app-c-autocomplete label", count: 1, text: "bar" do |label|
+      randomised_for = label[0]['for']
+    end
+    assert_select ".app-c-autocomplete input[name='foo'][id='#{randomised_for}']", count: 1
+  end
+
+  it "should render text input when id, name & label provided" do
+    render_component(name: "foo", label: { text: "bar" }, id: "baz")
+
+    assert_select ".app-c-autocomplete", count: 1
+    assert_select ".app-c-autocomplete label[for='baz']", count: 1, text: "bar"
+    assert_select ".app-c-autocomplete input[name='foo'][id='baz']", count: 1
+  end
+
+  it "renders datalist element associated with text input when options provided" do
+    render_component(
+      id: "basic-autocomplete",
+      name: "foo",
+      label: { text: "Countries" },
+      input: {
+        options: ["United Kingdom", "United States"]
+      }
+    )
+
+    assert_select ".govuk-label", text: "Countries", for: "basic-autocomplete"
+    assert_select ".app-c-autocomplete input[name='foo'][id='basic-autocomplete'][list='basic-autocomplete-list']", count: 1
+    assert_select "datalist#basic-autocomplete-list", count: 1
+    assert_select "datalist#basic-autocomplete-list option[value='United Kingdom']", count: 1
+    assert_select "datalist#basic-autocomplete-list option[value='United States']", count: 1
+  end
+
+  it "renders a default value if provided" do
+    render_component(
+      id: "foo",
+      name: "foo",
+      label: { text: "Countries" },
+      input: {
+        options: ["United Kingdom", "United States"],
+        value: "Spain"
```

Comment:
The default, which I assume is what `value` is. isn't one of the options, so doesn't that mean the default value isn't being rendered?

Reply:
Yes. In theory, the placeholder doesn't need to be one of the values in the options array (although it often is in practice).